### PR TITLE
service/dap: Add support for debug and test modes

### PR DESCRIPTION
**Pull request by polinasok**
This change also:
- refactors building helper functions from commands.go into its own gobuild package.
- makes processing untyped launch args (that are implementation-specific and not part of the DAP spec) more robust, verified by additional tests

### CODE REVIEW

File: service/dap/server.go [hunk 1]
```diff
@@ -16,12 +16,13 @@ import (
...
 	"path/filepath"
 
+	"github.com/go-delve/delve/pkg/gobuild"
 	"github.com/go-delve/delve/pkg/logflags"
 	"github.com/go-delve/delve/pkg/proc"
 	"github.com/go-delve/delve/service"
 	"github.com/go-delve/delve/service/api"
 	"github.com/go-delve/delve/service/debugger"
-	"github.com/google/go-dap"
+	"github.com/google/go-dap" // dap
```
Comment (eliben) [inline service/dap/server.go]: I'm not sure what's adding this, I just recently removed it - this comment is not needed

File: service/dap/server.go [hunk 2]
```diff
@@ -286,34 +292,71 @@ func (s *Server) onInitializeRequest(request *dap.InitializeRequest) {
...
 }
 
+// Output path for the compiled binary in debug or test modes.
+const debugBinary string = "./__debug_bin"
+
 func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
 	// TODO(polina): Respond with an error if debug session is in progress?
-	program, ok := request.Arguments["program"]
+
+	program, ok := request.Arguments["program"].(string)
```
Comment (eliben) [inline service/dap/server.go:301]: Sorry, I may be missing something here, but has the semantics of what's stored in `request.Arguments` changed?

Because before this change the `ok` was saying whether the key "program" is found in the map. After the change it's saying that the conversion to string is successful. Is this an intentional change?

File: service/dap/server_test.go [hunk 3]
```diff
@@ -172,14 +173,50 @@ func TestSetBreakpoint(t *testing.T) {
 	})
 }
 
-func TestBadLaunchRequests(t *testing.T) {
+func initLaunchRunDisconnect(t *testing.T, client *daptest.Client, launchRequest func()) {
```
Comment (eliben) [inline service/dap/server_test.go]: Document this function

File: pkg/gobuild/gobuild.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+// Remove the file at path and issue a warning to stderr if this fails.
+// This can be used to remove the temporary binary generated for the session.
+func Remove(path string) {
+	err := os.Remove(path)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "could not remove %v: %v\n", path, err)
+	}
+}
+
+// OptFlags generates default build flags to run off optimization and inlining.
```
Comment (eliben) [inline pkg/gobuild/gobuild.go]: "run off"? Did you mean "turn off" or something else?

File: service/dap/daptest/client.go [hunk 5]
```diff
@@ -137,12 +137,16 @@ func (c *Client) InitializeRequest() {
 }
 
 // LaunchRequest sends a 'launch' request.
-func (c *Client) LaunchRequest(program string, stopOnEntry bool) {
+// Takes arguments as interface{} to allow for testing of values
+// of unexpected types since the launch flags are implementation-specific
```
Comment (eliben) [inline service/dap/daptest/client.go]: This seems a bit concerning from a typing / usability point of view. Functions that take long lists of arguments are not great in the first place, but when all these arguments are untyped it's even worse. It becomes pretty hard to use the function correctly.

What is the actual goal here? Are you testing that integers instead of string paths produce expected errors? Since we don't have Go on the other side, the mapping between types and what goes into JSON isn't very clear anyway. From a cursory look at the other changes in this PR it's really hard to notice where you're actually using non-standard types.

Would it make sense to have a simple-to-use `LaunchRequest` for the common cases (which will take part in many tests, I imagine), and some special version to test a few corner cases of type handling in isolated cases?

**Review by eliben — COMMENTED** [review]
Made another pass :)

File: service/dap/server.go (hunk 1 above)
Comment (polinasok) [reply service/dap/server.go]: I believe Hana added this because the import ends with go-dap, but when you reference the items in the package, you prefix them with "dap".

File: pkg/gobuild/gobuild.go (hunk 4 above)
Comment (polinasok) [reply pkg/gobuild/gobuild.go]: Done. Odd typo.

File: service/dap/server.go (hunk 2 above)
Comment (polinasok) [reply service/dap/server.go:301]: I was missing a type check before, so if the program was not a string, then my old could would have panic'ed. This new code combines the old map check and the missing type check into one. If the item is in the map and is a string, then ok will be true. If the item is not in the map or is in the map, but is not a string, then ok will be false. https://play.golang.org/p/1oFdo4qNz4-

File: service/dap/server_test.go (hunk 3 above)
Comment (polinasok) [reply service/dap/server_test.go]: Renamed and added a comment.

File: service/dap/daptest/client.go (hunk 5 above)
Comment (polinasok) [reply service/dap/daptest/client.go]: I changed the code to use 2 versions of LaunchRequest. If you think I need different argument combinations, please let me know.

In general, LaunchRequest is the only struct where part of it won't be decoded with go types, but will be an untyped map. I added code in this CL to test not only for the presence of fields, but also for their type. And the new tests check that the code handles bad/missing values without panic.

File: service/dap/server.go (hunk 1 above)
Comment (eliben) [reply service/dap/server.go]: I don't think this comment is semantically necessary, though. One of my previous commits just removed it -- https://github.com/polinasok/delve/commit/23b8e59ab9837698221eb50fadcf6c5fcce2d1a7 -- so it'd be a shame to add it back.

It's perfectly fine in Go to have an import path "like-this" and referenced as "this", and many examples in Delve itself. E.g. see usage of `go-colorable` in https://github.com/go-delve/delve/blob/master/pkg/terminal/terminal_windows.go#L9

**Comment by eliben** [comment]
LGTM! Just one nit to remove the `// dap` comment from the import line

**Review by eliben — APPROVED** [review]

File: service/dap/server.go (hunk 1 above)
Comment (polinasok) [reply service/dap/server.go]: Removed. I have no attachment to this comment. I thought it was some kind of convention. It was still in my file as I copied it around trying to avoid conflicts with your commits.

**Comment by polinasok** [comment]
This is now https://github.com/go-delve/delve/pull/1901.
